### PR TITLE
[#4705] Ensure ChannelHandler.handlerAdded(...) is always called as f…

### PR DESCRIPTION
**Pull request by normanmaurer**
…irst method of the handler

Motivation:

If a user adds a ChannelHandler from outside the EventLoop it is possible to get into the situation that handlerAdded(...) is scheduled on the EventLoop and so called after another methods of the ChannelHandler as the EventLoop may already be executing on this point in time.

Modification:

- Ensure we always check if the handlerAdded(...) method was called already and if not add the currently needed call to the EventLoop so it will be picked up after handlerAdded(...) was called. This works as if the handler is added to the ChannelPipeline from outside the EventLoop the actual handlerAdded(...) operation is scheduled on the EventLoop.
- Some cleanup in the DefaultChannelPipeline

Result:

Correctly order of method executions of ChannelHandler.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler @trustin PTAL... this was quite tricky and I was not really come up with a good unit test :(

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 1]
```diff
@@ -257,23 +262,21 @@ public ChannelFuture write(Object msg) {
...
         return this;
     }
 
     @Override
     public ChannelFuture writeAndFlush(Object msg, ChannelPromise promise) {
         AbstractChannelHandlerContext next = findContextOutbound();
-        ChannelHandlerInvoker invoker = next.invoker();
-        invoker.invokeWrite(next, pipeline.touch(msg, next) , promise);
-        invoker.invokeFlush(next);
+        next.invokeWriteAndFlush(pipeline.touch(msg, next) , promise);
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: nit: kill space before `,`

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 2]
```diff
@@ -336,6 +339,291 @@ public boolean isRemoved() {
...
+                    assert handlerAdded;
+                    invoker.invokeWrite(AbstractChannelHandlerContext.this, msg, promise);
+                    invoker.invokeFlush(AbstractChannelHandlerContext.this);
+                }
+            });
+        }
+    }
+
+    private ChannelHandlerInvoker invoker() {
+        if (invoker == null) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: nit: `return invoker == null ? channel().unsafe().invoker() : invoker;`

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 3]
```diff
@@ -35,6 +36,18 @@
...
+    // to ensure we will never call another ChannelHandler method before handlerAdded(...) was called to guard against
+    // ordering issues. handlerAdded(...) MUST be the first method that is called per handler when become part of the
+    // ChannelPipeline in all cases. Not doing so may lead to unexpected side-effects as ChannelHandler implementations
+    // may need to do initialization steps before a ChannelHandler can be used.
+    //
+    // See:
+    //   - https://github.com/netty/netty/issues/4705
+    //
+    // No need to mark volatile as this will be made visible as next/prev is volatile.
+    private boolean handlerAdded;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Does this need to be a bit more generic to represent the implicit ordering dependencies between other methods? Are there other methods that could be called from other threads which may violate the order for other methods? For example does this need to be a `byte state` (or something similar) which each method needs to check if they can execute immediately or dispatch a task to the event loop to run later?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: no I think a boolean is fine here.

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: we only need to ensure we will delay until the handlerAdded was called

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 4]
```diff
@@ -336,6 +339,291 @@ public boolean isRemoved() {
...
+    private ChannelHandlerInvoker invoker() {
+        if (invoker == null) {
+            return channel().unsafe().invoker();
+        } else {
+            return invoker;
+        }
+    }
+
+    final void setHandlerAddedCalled() {
+        handlerAdded = true;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: does this need to be set back to false in handler removed?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: nope it should always stay true as it only signals that handlerAdded was called

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: sgtm ... was just informed that we are in the context :)

File: transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java [hunk 5]
```diff
@@ -96,13 +98,11 @@ public ChannelPipeline addFirst(String name, ChannelHandler handler) {
 
     @Override
     public synchronized ChannelPipeline addFirst(EventExecutorGroup group, String name, ChannelHandler handler) {
-        name = filterName(name, handler);
-        addFirst0(new DefaultChannelHandlerContext(this, findInvoker(group), name, handler));
-        return this;
+        return addFirst(findInvoker(group), name, handler);
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: why this change? won't we now have to go through 2 `synchronized(this)` blocks when this method is called?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: just to share more code... and I think it is honestly not a big deal in this case.

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: would it be OK to pull this out into another PR if it is not necessary for this PR?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: sure

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 6]
```diff
@@ -35,6 +36,18 @@
     private final String name;
     private boolean removed;
 
+    // This will set to true once the ChannelHandler.handlerAdded(...) method was called. We need to keep track of this
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: - Please use `/** ... */` because it's a field description.
- This will set to -> This is set to
- Please use `{@link ...}` where appropriate.

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 7]
```diff
@@ -35,6 +36,18 @@
...
+    // to ensure we will never call another ChannelHandler method before handlerAdded(...) was called to guard against
+    // ordering issues. handlerAdded(...) MUST be the first method that is called per handler when become part of the
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: - per handler -> for handler
- when become part of -> when it becomes a part of

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java (hunk 3 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Perhaps we could merge `removed` and `handlerAdded` into a single field?

**Comment by trustin** [comment]
Left a few nits. Looks good in overall.

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java (hunk 3 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: ok let me do this as int and bitshifting.

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: hmm maybe two boolean is still better .... Let us just keep it as it is. We can change later if we feel strong about it. WDYT ?

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 8]
```diff
@@ -35,6 +36,22 @@
...
+     * to guard againstordering issues. {@link ChannelHandler#handlerAdded(ChannelHandlerContext)} MUST be the first
+     * method that is called for handler when it becomes a part of the {@link ChannelPipeline} in all cases. Not doing
+     * so may lead to unexpected side-effects as {@link ChannelHandler} implementationsmay need to do initialization
+     * steps before a  {@link ChannelHandler} can be used.
+     *
+     * See <a href="https://github.com/netty/netty/issues/4705">#4705</a>
+     *
+     * No need to mark volatile as this will be made visible as next/prev is volatile.
+     */
+    private boolean handlerAdded;
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Another nit: we have `removed` field above. Should it be renamed to `handlerRemoved` or should this field be renamed to `added`?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Let me rebame to handlerRemoved

File: transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java [hunk 9]
```diff
@@ -159,10 +161,7 @@ public ChannelPipeline addBefore(String baseName, String name, ChannelHandler ha
     @Override
     public synchronized ChannelPipeline addBefore(
             EventExecutorGroup group, String baseName, String name, ChannelHandler handler) {
-        AbstractChannelHandlerContext ctx = getContextOrDie(baseName);
-        name = filterName(name, handler);
-        addBefore0(ctx, new DefaultChannelHandlerContext(this, findInvoker(group), name, handler));
-        return this;
+        return addBefore(findInvoker(group), baseName, name, handler);
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: Is this change necessary for this PR? If not can we revert this change and submit in a new PR?

File: transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java [hunk 10]
```diff
@@ -193,10 +192,7 @@ public ChannelPipeline addAfter(String baseName, String name, ChannelHandler han
     @Override
     public synchronized ChannelPipeline addAfter(
             EventExecutorGroup group, String baseName, String name, ChannelHandler handler) {
-        AbstractChannelHandlerContext ctx = getContextOrDie(baseName);
-        name = filterName(name, handler);
-        addAfter0(ctx, new DefaultChannelHandlerContext(this, findInvoker(group), name, handler));
-        return this;
+        return addAfter(findInvoker(group), baseName, name, handler);
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java]: Is this change necessary for this PR? If not can we revert this change and submit in a new PR?

**Comment by Scottmitch** [comment]
few comments and lgtm

**Comment by johnou** [comment]
ship ahoy?

**Comment by Scottmitch** [comment]
@normanmaurer - lgtm

**Comment by johnou** [comment]
@normanmaurer by chance do you plan on cutting 4.0.34 today too?

**Comment by normanmaurer** [comment]
@johnou no... on monday.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 as af39cb6b12d9c1a8cf97018dc038fbecc3ea1277
